File: 2016/day-1/part-1/main.cpp
#include <boost/algorithm/string.hpp>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

auto main() -> int
{
  std::ifstream file("input.txt");
  std::string instruction = "";
  std::vector<std::string> instructions = {};
  std::string pause = "";
  int x = 0;
  int y = 0;
  char heading = 'N';
  int steps = 0;
  std::string direction = "";

  while (std::getline(file, instruction, ','))
  {
    boost::trim(instruction);
    instructions.push_back(instruction);
  }

  for (auto instruction : instructions)
  {
    direction = instruction.substr(0, 1);
    steps = std::stoi(instruction.substr(1, instruction.length()));

    if (direction == "R")
    {
      if (heading == 'N')
      {
        heading = 'E';
        x += steps;
      }
      else if (heading == 'E')
      {
        heading = 'S';
        y -= steps;
      }
      else if (heading == 'S')
      {
        heading = 'W';
        x -= steps;
      }
      else if (heading == 'W')
      {
        heading = 'N';
        y += steps;
      }
    }
    else if (direction == "L")
    {
      if (heading == 'N')
      {
        heading = 'W';
        x -= steps;
      }
      else if (heading == 'W')
      {
        heading = 'S';
        y -= steps;
      }
      else if (heading == 'S')
      {
        heading = 'E';
        x += steps;
      }
      else if (heading == 'E')
      {
        heading = 'N';
        y += steps;
      }
    }
  }

  std::cout << "Final position: " << x << ", " << y << std::endl;
  std::cout << "Distance: " << abs(x) + abs(y) << std::endl;
  std::cin >> pause;

  return 0;
}

File: 2016/day-1/part-1/src/main.rs
use std::fs;

enum Direction {
    North,
    East,
    South,
    West,
}

struct Point {
    x: i32,
    y: i32,
    direction: Direction,
}

fn main() {
    let mut position = Point {
        x: 0,
        y: 0,
        direction: Direction::North,
    };
    let instructions = fs::read_to_string("input.txt").expect("Error reading input.txt");

    for token in instructions.split(",").map(|s| s.trim()) {
        let steps: i32 = token[1..].parse().expect("Error parsing steps");
        match token.chars().nth(0).unwrap() {
            'R' => match position.direction {
                Direction::North => position.direction = Direction::East,
                Direction::East => position.direction = Direction::South,
                Direction::South => position.direction = Direction::West,
                Direction::West => position.direction = Direction::North,
            },
            'L' => match position.direction {
                Direction::North => position.direction = Direction::West,
                Direction::East => position.direction = Direction::North,
                Direction::South => position.direction = Direction::East,
                Direction::West => position.direction = Direction::South,
            },
            _ => panic!("Invalid instruction"),
        }
        match position.direction {
            Direction::North => position.y += steps,
            Direction::East => position.x += steps,
            Direction::South => position.y -= steps,
            Direction::West => position.x -= steps,
        }
    }
    println!("Distance: {}", position.x.abs() + position.y.abs());
}
